[TRL] Liger-GRPO blog (#2864)
huggingface/blog: liger-grpo.md

Here's the visualization of the optimization (ref: [Byron Hsu](https://x.com/hsu_byron/status/1866577403918917655)):

![liger-chunked-loss](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/liger-grpo/image7.gif)

## Plug-and-Play integration with TRL

We recently integrated Liger GRPO with TRL in PR [#3184](https://github.com/huggingface/trl/pull/3184), so now you can use the Liger GRPO loss just by setting `use_liger_loss` to `True` in your `GRPOConfig` and enjoy the memory savings!

Heads up: these features aren't in the latest TRL release yet, so you'll need to install TRL from source for now:

```bash
pip install "trl[liger] @ git+https://github.com/huggingface/trl.git"
```

and then you can use it like this:
```python
from trl import GRPOConfig, GRPOTrainer

> 

Reviewer: ```suggestion
def reward_len(completions, **kwargs):
    return [-abs(20 - len(completion)) for completion in completions]
```

> 

Reviewer: ```suggestion
train_dataset = load_dataset("trl-lib/tldr", split="train")
```

training_args = GRPOConfig(output_dir="Qwen3-0.6B-GRPO", use_liger_loss=True)

trainer = GRPOTrainer(
    model="Qwen/Qwen3-0.6B-Instruct",

>     reward_func=reward_len,

Reviewer: ```suggestion
    reward_funcs=reward_len,
```

Here, we show a multi-GPU GRPO training plot using FSDP and PEFT, where we compare the maximum training batch size possible with and without the Liger Loss across different Qwen3 model sizes. We found that with Liger, we were able to bump up the batch size by around **1.5 to 1.8x**!

![peft-batch-size-vs-model-size](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/liger-grpo/image6.png)

> 

Reviewer: ```suggestion
## Scaling even further with vLLM
Liger loss can also be used alongside generation with implemented vllm-server in TRL for significantly faster text generation. This enables efficient rollout collection during training with minimal overhead and seamless integration. 
All you need to do is the following;
```

```suggestion
first run the server using: 
`CUDA_VISIBLE_DEVICES=1 trl vllm-serve --model "Qwen/Qwen3-0.6B"`

then the training script being like;

```python
from trl import GRPOConfig, GRPOTrainer
from datasets import load_dataset

training_args = GRPOConfig(output_dir="Qwen3-0.6B-GRPO", use_liger_loss=True, use_vllm=True, logging_steps=10)

def reward_len(completions, **kwargs):
    return [-abs(20 - len(completion)) for completion in completions]

dataset = load_dataset("trl-lib/tldr", split="train[:1%]")

trainer = GRPOTrainer(
    model="Qwen/Qwen3-0.6B",
    reward_funcs=reward_len,
    args=training_args,
    train_dataset=dataset,
)

trainer.train()
```
then run it using:

`CUDA_VISIBLE_DEVICES=0 accelerate train.py`
```

---
title: "🐯 Liger GRPO meets TRL" 
thumbnail: /blog/assets/liger-grpo/thumbnail.png
authors:
- user: shisahni
  guest: true
  org: LinkedIn
- user: kashif
- user: smohammadi
  guest: true
  org: axolotl-ai-co
- user: ShirinYamani
- user: m0m0chen
  guest: true
  org: LinkedIn
- user: liberty4321
  guest: true
  org: LinkedIn
---

# 🐯 Liger GRPO meets TRL

TL; DR

> [Liger](https://github.com/linkedin/Liger-Kernel) supercharges [TRL](https://github.com/huggingface/trl)’s [GRPO Trainer](https://huggingface.co/docs/trl/grpo_trainer) by slashing memory usage by **40%** with zero drop in model quality. We also added support for **FSDP** and **PEFT**, making it easier than ever to scale GRPO across multiple GPUs.

Author: ```suggestion
[Liger](https://github.com/linkedin/Liger-Kernel) supercharges [TRL](https://github.com/huggingface/trl)’s Group Relative Policy Optimization [GRPO Trainer](https://huggingface.co/docs/trl/grpo_trainer) by slashing memory usage by **40%** with zero drop in model quality. We also added support for **FSDP** and **PEFT**, making it easier than ever to scale GRPO across multiple GPUs.
```

---
title: "🐯 Liger GRPO meets TRL" 
thumbnail: /blog/assets/liger-grpo/thumbnail.png
authors:
- user: shisahni
  guest: true
- user: kashif
- user: smohammadi

>   guest: true
> 

Reviewer 2: ```suggestion
  guest: true
  org: axolotl-ai-co
```

---

# 🐯 Liger GRPO meets TRL

TL; DR

> [Liger](https://github.com/linkedin/Liger-Kernel) supercharged [TRL](https://github.com/huggingface/trl)’s [GRPO Trainer](https://huggingface.co/docs/trl/grpo_trainer) by slashing memory usage by **40%** with zero drop in model quality. We also added support for **FSDP** and **PEFT**, making it easier than ever to scale GRPO across multiple GPUs. 

Reviewer 3: turbo nit:
```suggestion
[Liger](https://github.com/linkedin/Liger-Kernel) supercharges [TRL](https://github.com/huggingface/trl)’s [GRPO Trainer](https://huggingface.co/docs/trl/grpo_trainer) by slashing memory usage by **40%** with zero drop in model quality. We also added support for **FSDP** and **PEFT**, making it easier than ever to scale GRPO across multiple GPUs. 
```

> ## Introduction & Motivation

Reviewer 3: nit:
```suggestion
## Motivation
```

> RLHF (Reinforcement Learning from Human Feedback) has been an effective way to get models to behave the way we want them to in real-world scenarios. Traditionally, though, doing RLHF has meant juggling a bunch of different models, i.e., an actor, critic, reward model, and reference model, all working together to “teach” the main model how to act. But this setup comes with the overhead of building and maintaining a pretty complex infrastructure to keep everything running smoothly and make sure all the data flows between them properly.

Reviewer 3: RLHF is not an algorithm per se, so I suggest reframing this in terms of RL being applied to optimise LLMs more generally, with PPO being the algorithm of choice until recently.

> RLHF (Reinforcement Learning from Human Feedback) has been an effective way to get models to behave the way we want them to in real-world scenarios. Traditionally, though, doing RLHF has meant juggling a bunch of different models, i.e., an actor, critic, reward model, and reference model, all working together to “teach” the main model how to act. But this setup comes with the overhead of building and maintaining a pretty complex infrastructure to keep everything running smoothly and make sure all the data flows between them properly.
> 
> More recently, DeepSeek’s R1 helped popularize GRPO, which simplifies things quite a bit without losing performance. GRPO gets rid of the critic model and instead uses the average reward of sampled outputs produced in response to the same prompt as the baseline. The following diagram shows GRPO vs PPO training pipeline (ref: [DeepSeekMath: Pushing the Limits of Mathematical Reasoning in Open Language Models](https://huggingface.co/papers/2402.03300)):

Reviewer 4: ```suggestion
Fine-tuning language models using reinforcement learning (RL) is a crucial step in a model's training lifecycle for steering models towards desirable behaviours which are more complex than can be achieved through typical supervised fine-tuning. Traditional RL applications have used Reinforcement Learning from Human Feedback (RLHF) with the Proximal Policy Optimization (PPO) algorithm, which uses a separately trained reward model to steer the behaviour of our model which is being fine-tuned. 

However, RLHF with PPO is a very resource-hungry approach - typical applications of RLHF involve loading multiple models in memory (policy, value, reward, and reference models), and require several iterations of fine-tuning reward and base models to achieve the desired results. The success of RLHF also depends on the capability of the reward model to effectively discriminate between desired and un-desired behaviour from our model.

Group Relative Policy Optimization (GRPO) has seen significant recent popularity alongside DeepSeek's R1 model. GRPO eschews the pre-trained reward model and value models used in RLHF and instead relies on *verifiable reward functions* which can check the correctness of a model's output in a closed-form manner i.e. without an external reward model. This has resulted in massive improvements when using GRPO instead of PPO for fine-tuning on domains which are easily verifiable, such as teaching a model to reason, and perform well on math and coding tasks. 

The following diagram shows GRPO vs PPO training pipeline (ref: [DeepSeekMath: Pushing the Limits of Mathematical Reasoning in Open Language Models](https://huggingface.co/papers/2402.03300)):
```

Reviewer 4: Oh I totally missed that. We could update the text, or use a different diagram as verifiable rewards are a much more common use case for GRPO

![PPO-vs-GRPO](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/liger-grpo/image5.png)

> That said, RL training still eats up a ton of GPU memory, so there’s still plenty of room for optimizations here. In this blog post, we talk about an optimization that we recently added to TRL that cuts peak memory usage by 30% during GRPO Training, and we also dive into how to scale GRPO to multiple GPUs and nodes without losing performance or correctness.

Reviewer 3: Align with the TL;DR value (or lower the TL;DR value)
```suggestion
That said, RL training still eats up a ton of GPU memory, so there’s still plenty of room for optimizations here. In this blog post, we talk about an optimization that we recently added to TRL that cuts peak memory usage by 40% during GRPO Training, and we also dive into how to scale GRPO to multiple GPUs and nodes without losing performance or correctness.
```

## How Liger Kernel slashes memory for GRPO

We extended the Liger Chunked Loss approach to GRPO Loss, which lets us avoid having to store the full logits in memory for every training step. We do this by chunking the input to the `lm_head` across the batch and running the forward pass one chunk at a time.

But if you just implement it in a straightforward way, you won’t actually be able to reduce the peak memory since you’d still need to keep all the logits in GPU memory for the backward pass. To get around that, we calculate the gradients for each loss chunk (with respect to the `input chunk` and the `lm_head weights`) during the forward pass, and then accumulate them as we go through each chunk.

Here’s the visualization of the optimization:

![liger-chunked-loss](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/liger-grpo/image7.gif)

## Plug-and-Play integration with TRL

We recently integrated Liger GRPO with TRL in PR [#3184](https://github.com/huggingface/trl/pull/3184), so now you can use the Liger GRPO loss just by setting `use_liger_loss` to `True` in your `GRPOConfig` and enjoy the memory savings!

> Heads up: these features aren’t in the latest TRL release yet, so you’ll need to install TRL from source for now.

Reviewer 3: Maybe add a command to pip install from main?

Here’s the plots of peak memory usage vs batch size for both FP32 and BF16 training. As expected, the memory savings get better with larger batch sizes since we chunk along the batch dimension. So when the batch size goes up, the Liger chunked loss ends up using a lot less memory, up to 40% less, compared to the regular (non-liger) version. 

Quick note: Right now, we only support FP32, but we're working on open-sourcing BF16 support for Liger GRPO in TRL. The BF16 results shown here are from internal patches we’ve been testing.

![Mem-vs-batch-size-fp32](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/liger-grpo/image3.png)

![Mem-vs-batch-size-bf16](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/liger-grpo/image4.png)

We also show that Liger Loss is effectively accurate. As seen in the plot, rewards over training steps stay pretty much the same as what you’d see using the standard TRL implementation.

> ![reward-vs-step](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/liger-grpo/image1.png)

Reviewer 2: Makes sense, thank you!

---
title: "🐯 Liger GRPO meets TRL" 
thumbnail: /blog/assets/liger-grpo/thumbnail.png
authors:
- user: shisahni
  guest: true
  org: LinkedIn
- user: kashif
- user: smohammadi
  guest: true
  org: axolotl-ai-co
---

# 🐯 Liger GRPO meets TRL

TL; DR
[Liger](https://github.com/linkedin/Liger-Kernel) supercharges [TRL](https://github.com/huggingface/trl)’s [GRPO Trainer](https://huggingface.co/docs/trl/grpo_trainer) by slashing memory usage by **40%** with zero drop in model quality. We also added support for **FSDP** and **PEFT**, making it easier than ever to scale GRPO across multiple GPUs. 

## Motivation

> Fine-tuning language models using reinforcement learning (RL) is a crucial step in a model's training lifecycle for steering models towards desirable behaviours which are more complex than can be achieved through typical supervised fine-tuning. Reinforcement learning (RL) has traditionally been applied to optimize large language models (LLMs) using the Proximal Policy Optimization (PPO) algorithm. This approach, often associated with Reinforcement Learning from Human Feedback (RLHF), utilizes a separately trained reward model to guide the fine-tuning of the primary model. 

Reviewer 4: ```suggestion
Fine-tuning language models using reinforcement learning (RL) is a crucial step in a model's training lifecycle for steering models towards desirable behaviours which are more complex than can be achieved through typical supervised fine-tuning. RL has traditionally been applied to optimize large language models (LLMs) using the Proximal Policy Optimization (PPO) algorithm. This approach, often associated with Reinforcement Learning from Human Feedback (RLHF), utilizes a separately trained reward model to guide the fine-tuning of the primary model. 
```
double acronym defn

But if you just implement it in a straightforward way, you won't actually be able to reduce the peak memory since you'd still need to keep all the logits in GPU memory for the backward pass. To get around that, we calculate the gradients for each loss chunk (with respect to the `input` chunk and the `lm_head` weight`) during the forward pass, and then accumulate them as we go through each chunk.

Here's the visualization of the optimization (ref: [Byron Hsu](https://x.com/hsu_byron/status/1866577403918917655)):

![liger-chunked-loss](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/liger-grpo/image7.gif)

## Plug-and-Play integration with TRL

We recently integrated Liger GRPO with TRL in PR [#3184](https://github.com/huggingface/trl/pull/3184), so now you can use the Liger GRPO loss just by setting `use_liger_loss` to `True` in your `GRPOConfig` and enjoy the memory savings!

> Heads up: these features aren't in the latest TRL release yet, so you'll need to install TRL from source for now:

Reviewer 5: We could release v0.18 very soon, wdyt @kashif?

Author: true...
